src/linker/Linker.Dataflow/ValueNode.cs

```diff
@@ -0,0 +1,1258 @@
+		}
+
+		public override bool Equals (ValueNode other)
+		{
+			if (other == null)
+				return false;
+			if (this.Kind != other.Kind)
+				return false;
+
+			ConstIntValue otherCiv = (ConstIntValue)other;
+			return Value == otherCiv.Value;
+		}
+
+		protected override string NodeToString ()
+		{
+			return ValueNodeDump.ValueNodeToString (this, Value);
+		}
+	}
+
+	class ArrayValue : ValueNode
+	{
+		protected override int NumChildren => 1;
+
+		/// <summary>
+		/// Constructs an array value of the given size
+		/// </summary>
+		public ArrayValue (ValueNode size)
+		{
+			Kind = ValueNodeKind.Array;
+			Size = size ?? UnknownValue.Instance;
+		}
+
+		public ValueNode Size { get; }
+
+		public override int GetHashCode ()
+		{
+			return HashUtils.CalcHashCode (Kind, Size);
+		}
+
+		public override bool Equals (ValueNode other)
+		{
+			if (other == null)
+				return false;
+			if (this.Kind != other.Kind)
+				return false;
+
+			ArrayValue otherArr = (ArrayValue)other;
+			return Size.Equals (otherArr.Size);
+		}
+
+		protected override string NodeToString ()
+		{
+			return ValueNodeDump.ValueNodeToString (this, Size);
+		}
+
+		protected override IEnumerable<ValueNode> EvaluateUniqueValues ()
+		{
+			foreach (var sizeConst in Size.UniqueValues)
+				yield return new ArrayValue (sizeConst);
+		}
+
+		protected override ValueNode ChildAt (int index)
+		{
+			if (index == 0) return Size;
+			throw new InvalidOperationException ();
+		}
+	}
+
+	#region ValueNode Collections
+	public class ValueNodeList : List<ValueNode>
+	{
+		public ValueNodeList ()
+		{
+		}
+
+		public ValueNodeList (int capacity)
+			: base (capacity)
+		{
+		}
+
+		public ValueNodeList (List<ValueNode> other)
+			: base (other)
+		{
+		}
+
+		public override int GetHashCode ()
+		{
+			return HashUtils.CalcHashCodeEnumerable (this);
+		}
+
+		public override bool Equals (object other)
+		{
+			ValueNodeList otherList = other as ValueNodeList;
+			if (otherList == null)
+				return false;
+
+			if (otherList.Count != Count)
+				return false;
+
+			for (int i = 0; i < Count; i++) {
+				if (!otherList [i].Equals (this [i]))
+					return false;
+			}
+			return true;
+		}
+	}
+
+	class ValueNodeHashSet : HashSet<ValueNode>
+	{
+		public override int GetHashCode ()
+		{
+			return HashUtils.CalcHashCodeEnumerable (this);
+		}
+
+		public override bool Equals (object other)
+		{
+			ValueNodeHashSet otherSet = other as ValueNodeHashSet;
+			if (otherSet == null)
+				return false;
+
+			if (otherSet.Count != Count)
+				return false;
+
+			IEnumerator<ValueNode> thisEnumerator = this.GetEnumerator ();
+			IEnumerator<ValueNode> otherEnumerator = otherSet.GetEnumerator ();
+
+			for (int i = 0; i < Count; i++) {
+				thisEnumerator.MoveNext ();
+				otherEnumerator.MoveNext ();
+				if (!thisEnumerator.Current.Equals (otherEnumerator.Current))
+					return false;
+			}
+			return true;
+		}
+	}
+	#endregion
+
+	static class HashUtils
```

Comment:
Why is this needed when there is HashCode.Combine ?

Reply:
HashCode is not available on net472.

Reply:
There is https://www.nuget.org/packages/Microsoft.Bcl.HashCode for such cases